crates/polars-io/src/partition.rs

```diff
@@ -139,100 +140,155 @@ pub fn write_partitioned_dataset<S>(
 
-            file_write_options
-                .to_writer(f)
-                .finish(&mut part_df.slice(slice_start as i64, rows_per_file))?;
-        }
-    }
+    let finish_part_df = |df: DataFrame| {
+        let dir_path = init_part_base_dir(&df)?;
+        let (n_files, rows_per_file) = get_n_files_and_rows_per_file(&df);
 
-    Ok(())
-}
+        if n_files == 1 {
+            write_part(df.clone(), &dir_path.join(get_path_for_index(0)))
+        } else {
+            let mut parts = UnitVec::new();
 
-/// Creates an iterator of (hive partition path, DataFrame) pairs, e.g.:
-/// ("a=1/b=1", DataFrame)
-fn get_hive_partitions_iter<'a, S>(
-    df: &'a DataFrame,
-    partition_by: &'a [S],
-) -> PolarsResult<Box<dyn Iterator<Item = (String, DataFrame)> + 'a>>
-where
-    S: AsRef<str>,
-{
-    let schema = df.schema();
-
-    let partition_by_col_idx = partition_by
-        .iter()
-        .map(|x| {
-            let Some(i) = schema.index_of(x.as_ref()) else {
-                polars_bail!(ColumnNotFound: "{}", x.as_ref())
-            };
-            Ok(i)
-        })
-        .collect::<PolarsResult<Vec<_>>>()?;
-
-    let get_hive_path_part = move |df: &DataFrame| {
-        const CHAR_SET: &percent_encoding::AsciiSet = &percent_encoding::CONTROLS
-            .add(b'/')
-            .add(b'=')
-            .add(b':')
-            .add(b' ');
+            for v in (0..df.height()).step_by(rows_per_file).enumerate() {
+                parts.push(v);
+            }
 
-        let cols = df.get_columns();
+            parts
+                .chunks(MAX_OPEN_FILES)
+                .map(|chunk| {
+                    chunk
+                        .into_par_iter()
+                        .map(|&(idx, slice_start)| {
+                            let df = df.slice(slice_start as i64, rows_per_file);
+                            write_part(df.clone(), &dir_path.join(get_path_for_index(idx)))
+                        })
+                        .reduce(
+                            || PolarsResult::Ok(()),
+                            |a, b| if a.is_err() { a } else { b },
+                        )
+                })
+                .reduce(|a, b| if a.is_err() { a } else { b })
+                .unwrap_or(PolarsResult::Ok(()))
+        }
+    };
 
-        partition_by_col_idx
-            .iter()
-            .map(|&i| {
-                let s = &cols[i].slice(0, 1).cast(&DataType::String).unwrap();
-
-                format!(
-                    "{}={}",
-                    s.name(),
-                    percent_encoding::percent_encode(
-                        s.str()
-                            .unwrap()
-                            .get(0)
-                            .unwrap_or("__HIVE_DEFAULT_PARTITION__")
-                            .as_bytes(),
-                        CHAR_SET
+    POOL.install(|| match groups {
+        GroupsProxy::Idx(idx) => idx
+            .all()
+            .chunks(MAX_OPEN_FILES)
```

Comment:
We can't do any semaphore logic because of Rayon's work-stealing, so I've just applied the limit by batching the groups instead.